### PR TITLE
Update supported Python versions and dependencies

### PR DESCRIPTION
**Pull request by ned9**
Assorted changes to add support for Python 3.10, 3.11, and 3.12

* remove support for Python 3.7 (it's end-of-life) and 3.8 (Numpy no longer supports it)
* add support for Python 3.10, 3.11, and 3.12
* remove all constraints for dependencies in `pyproject.toml`
* update `poetry.lock` with updated dependencies, to get latest versions of all deps
* small change to test to handle Pandas API change: keyword argument renamed in `pandas.testing.assert_frame_equal`

Important call-out is probably the relaxing of all version constraints in `pyproject.toml` dependencies. I've never understood why Poetry defaults to applying a `^` constraint all the way to the patch version number for the version of the package installed when you run `poetry add`, given that the resolved versions are pinned in `poetry.lock`. In my experience this tends to increase the chances of dependency hell down the line. If incompatibilities or bugs are encountered in specific dependency versions, we can add them when (and if) they are encountered.

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -1,35 +1,35 @@
...
+mypy = "*"
+pytest-mock = "*"
 
 [build-system]
 requires = ["poetry-core>=1.0.0"]
 build-backend = "poetry.core.masonry.api"
 
 [tool.mypy]
-python_version = "3.7"
+python_version = "3.8"
```
Comment (buzzysaari) [inline pyproject.toml:27]: You should also update the versions mentioned in README

File: tests/test_data_loader.py [hunk 2]
```diff
@@ -39,4 +39,4 @@ def test_preprocess_data(mocker, mock_df: pd.DataFrame):
         index=[0, 1, 2],
     )
     expected_df.index.name = "id"
-    assert_frame_equal(data_loader.party_data, expected_df, check_less_precise=3)
+    assert_frame_equal(data_loader.party_data, expected_df, rtol=3)
```
Comment (buzzysaari) [inline tests/test_data_loader.py:42]: The test failed for me:

>       assert_frame_equal(data_loader.party_data, expected_df, rtol=3)
E       AssertionError: DataFrame are different
E       
E       DataFrame shape mismatch
E       [left]:  (4, 6)
E       [right]: (3, 3)

Comment (buzzysaari) [inline tests/test_data_loader.py:42]: The test failed for me, but is that expected?

Comment (ned9) [reply tests/test_data_loader.py:42]: yep, that's expected. this test will only pass if the pre-processing exercise has been completed successfully. the test passes for me with a local solution

File: pyproject.toml (hunk 1 above)
Comment (ned9) [reply pyproject.toml:27]: ah yes, good catch! I also made some other minor README improvements while I was in there.

**Pull request merged**